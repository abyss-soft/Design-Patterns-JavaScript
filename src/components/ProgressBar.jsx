import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { getAnswers, getRemaining } from '../selectors';

const Container = styled.div`
  display: flex;
  height: 3px;
  justify-content: space-between;
  margin: 1.25rem 0 1rem;
  width: 100%;
`;

const Step = styled.span`
  background: ${props => (props.nature ? props.theme[props.nature] : props.theme.ALTO)};
  display: flex;
  height: 2px;
  width: 3.75%;
`;

export const ProgressBar = props => {
  const { answers, remaining } = props;

  return (
    <Container>
      {answers.map(({ uuid, answered, correct }) => {
        let nature;

        if (answered && correct) {
          nature = 'ATLANTIS';
        } else {
          nature = 'CRIMSON';
        }

        return <Step key={uuid} nature={nature} />;
      })}

      {remaining.map(({ uuid }) => (
        <Step key={uuid} />
      ))}
    </Container>
  );
};

ProgressBar.propTypes = {
  answers: PropTypes.array.isRequired,
  remaining: PropTypes.array.isRequired
};

export default connect(state => ({
  answers: getAnswers(state),
  remaining: getRemaining(state)
}))(ProgressBar);
