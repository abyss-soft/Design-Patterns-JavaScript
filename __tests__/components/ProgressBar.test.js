import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import 'jest-styled-components';
import { ThemeProvider } from 'styled-components';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import { themeCommon } from '../../src/styles/themes/theme.common';
import ProgressBar from '../../src/components/ProgressBar';

const mockStore = configureMockStore();
const store = mockStore({
  progress: {
    answers: [
      { uuid: 1, answered: true, correct: true },
      { uuid: 2, answered: true, correct: true },
      { uuid: 3, answered: true, correct: true },
      { uuid: 4, answered: true, correct: false }
    ],
    remaining: [
      { uuid: 5, answered: false, correct: null },
      { uuid: 6, answered: false, correct: null },
      { uuid: 7, answered: false, correct: null },
      { uuid: 8, answered: false, correct: null }
    ]
  }
});

describe('<ProgressBar /> component', () => {
  let result;

  beforeEach(() => {
    // the wrapper DIV added to get rid of iterator key warning
    result = shallow(
      <div key="unique">
        <Provider store={store}>
          <ThemeProvider theme={themeCommon}>
            <ProgressBar />
          </ThemeProvider>
        </Provider>
      </div>
    );
  });

  it('renders without crashing', () => {
    expect(result).toBeTruthy();
  });

  it('shows the number of steps', () => {
    const tree = renderer.create(result).toJSON();
    expect(tree.children[0].children.length).toBe(8);
  });

  it('has styled-component rendered classes', () => {
    const tree = renderer.create(result).toJSON();
    expect(tree.children[0].props.className).toBeDefined();
  });

  it('matches the snapshot', () => {
    const tree = renderer.create(result).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
