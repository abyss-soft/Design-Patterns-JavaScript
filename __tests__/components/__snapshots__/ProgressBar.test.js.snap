// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressBar /> component matches the snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 3px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin: 1.25rem 0 1rem;
  width: 100%;
}

.c1 {
  background: #9ACD32;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2px;
  width: 3.75%;
}

.c2 {
  background: #E22A23;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2px;
  width: 3.75%;
}

.c3 {
  background: #D8D8D8;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2px;
  width: 3.75%;
}

<div>
  <div
    className="c0"
  >
    <span
      className="c1"
    />
    <span
      className="c1"
    />
    <span
      className="c1"
    />
    <span
      className="c2"
    />
    <span
      className="c3"
    />
    <span
      className="c3"
    />
    <span
      className="c3"
    />
    <span
      className="c3"
    />
  </div>
</div>
`;
